### PR TITLE
✨ feat(VSecM Safe): 460 manual root key update k8s secret option

### PR DESCRIPTION
**Pull request by gurkanguray**
# Manual Root Key Updates K8s Secret

## Description

VSECM_MANUAL_ROOT_KEY_UPDATES_K8S_SECRET env variable added for giving an option to updating internal k8s secrets when manual root key provided

## Changes

- Renamed `app/safe/internalprivates.go` to `app/safe/internal/bootstrap/persist.go` since we use this function in the `app/safe/internal/server/route/receive.go` to persist keys with the manual root key. Therefore, it's not private to `bootstrap` anymore.
- Added `VSECM_MANUAL_ROOT_KEY_UPDATES_K8S_SECRET` to `core/safe.go` with its unit tests.
- Although they belong to the v0.23.0, I assumed that this is the main branch and the changes should reflected in the next version, so I added the `VSECM_MANUAL_ROOT_KEY_UPDATES_K8S_SECRET` env variable to the following yaml files:
  - k8s/0.23.0/local/vsecm-distroless.yaml
  - k8s/0.23.0/local/vsecm-distroless-fips.yaml
  - k8s/0.23.0/local/vsecm-photon.yaml
  - k8s/0.23.0/local/vsecm-photon-fips.yaml
  - k8s/0.23.0/remote/vsecm-distroless.yaml
  - k8s/0.23.0/remote/vsecm-distroless-fips.yaml
  - k8s/0.23.0/remote/vsecm-photon-fips.yaml
  - helm-charts/0.23.0/charts/safe/values.yaml
- Added `VSECM_MANUAL_ROOT_KEY_UPDATES_K8S_SECRET` information details to docs/_pages/0110-configuration.md file

## Test Policy Compliance

- [x] I have added or updated unit tests for my changes.
- [x] I have included integration tests where applicable.
- [x] All new and existing tests pass successfully.

## Code Quality

- [x] I have followed the coding standards for this project.
- [x] I have performed a self-review of my code.
- [x] My code is well-commented, particularly in areas that may be difficult 
      to understand.

## Documentation

- [x] I have made corresponding changes to the documentation (if applicable).
- [x] I have updated any relevant READMEs or wiki pages.

## Checklist

Before you submit this PR, please make sure:
- [x] You have read [the contributing guidelines][contributing] and 
      *especially* the [test policy][test-policy].
- [x] You have thoroughly tested your changes.
- [x] You have followed all the contributing guidelines for this project.
- [x] You understand and agree that your contributions will be publicly available 
  under the project’s license.

[contributing]: https://vsecm.com/docs/contributing/
[test-policy]: https://vsecm.com/docs/contributing/#add-tests-for-new-features

*By submitting this pull request, you confirm that my contribution is made under 
the terms of the project’s license and that you have the authority to grant 
these rights.*

---

Thank you for your contribution to [**VMware Secrets Manager**](https://vsecm.com)
🐢⚡️!

### CODE REVIEW

File: docs/_pages/0110-configuration.md [hunk 1]
```diff
@@ -22,24 +22,24 @@ next_url: /docs/use-the-source/
```
Comment (v0lkan) [inline docs/_pages/0110-configuration.md:22]: There is IDE-generated auto-formatting here.

Can we please revert unrelated changes?

File: examples/federation-workshop/cluster-2/safe/Deployment.yaml [hunk 2]
```diff
@@ -72,6 +72,8 @@ spec:
               value: "/key/key.txt"
             - name: VSECM_ROOT_KEY_INPUT_MODE_MANUAL
               value: "false"
+            - name: VSECM_MANUAL_ROOT_KEY_UPDATES_K8S_SECRET
+              value: "false"
```
Comment (v0lkan) [inline examples/federation-workshop/cluster-2/safe/Deployment.yaml:76]: Why do we have these in example workload manifests?

`VSECM_MANUAL_ROOT_KEY_UPDATES_K8S_SECRET` is for **VSecM Safe** to use; it will be meaningless for any other workload.

File: app/safe/internal/bootstrap/bootstrap.go [hunk 3]
```diff
@@ -188,7 +188,7 @@ func CreateCryptoKey(id *string, updatedSecret chan<- bool) {
 
 	log.InfoLn(id, "Generated public key, private key, and aes seed")
 
-	if err = persistKeys(privateKey, publicKey, aesSeed); err != nil {
+	if err = PersistKeys(privateKey, publicKey, aesSeed); err != nil {
```
Comment (v0lkan) [inline app/safe/internal/bootstrap/bootstrap.go:191]: What’s the reasoning behind making this method public?

From the looks of it, it’s only used in this module.

Is there a specific need to keep it public?

Comment (gurkanguray) [reply app/safe/internal/bootstrap/bootstrap.go:191]: [ReceiveKeys()](https://github.com/vmware-tanzu/secrets-manager/pull/564/files/3dcc0f8086cf86e2185778adacdc4c4206f6df42#diff-722ad5277a6975ab286b63b74eddb3809114bd3efc157ebb33a2160f48b6f0d6R27) function of the Safe saves the input keys and I aim to perform the [persistKeys()](https://github.com/vmware-tanzu/secrets-manager/pull/564/files/3dcc0f8086cf86e2185778adacdc4c4206f6df42#diff-722ad5277a6975ab286b63b74eddb3809114bd3efc157ebb33a2160f48b6f0d6R71) operations in that function if the relative env variable is set. To avoid duplication, I planned to export the private persistKeys() from the bootstrap module to call it from the /server/route/receive.go file.

File: examples/federation-workshop/cluster-2/safe/Deployment.yaml (hunk 2 above)
Comment (gurkanguray) [reply examples/federation-workshop/cluster-2/safe/Deployment.yaml:76]: Right! I'm going to remove it. Thank you

File: app/safe/internal/bootstrap/bootstrap.go (hunk 3 above)
Comment (v0lkan) [reply app/safe/internal/bootstrap/bootstrap.go:191]: Noted. Fair enough. — Let’s see how it unfolds :)

File: app/safe/internal/server/route/receive.go [hunk 4]
```diff
@@ -65,6 +67,12 @@ func ReceiveKeys(cid string, w http.ResponseWriter, r *http.Request, spiffeid st
 	keysCombined := agePrivateKey + "\n" + agePublicKey + "\n" + aesCipherKey
 	state.SetRootKey(keysCombined)
 
+	if env.RootKeyInputModeManual() && env.ManualRootKeyUpdatesKubernetesSecret() {
+		if err := bootstrap.PersistKeys(agePrivateKey, agePublicKey, aesCipherKey); err != nil {
+			log.FatalLn(&cid, "Failed to persist keys", err.Error())
```
Comment (v0lkan) [inline app/safe/internal/server/route/receive.go:72]: Let’s not crash, but just warn or something.

My worry is: if safe is not working with a backing store, killing it will lose whatever is in its memory.

`log.FatalLn` kills the process by design. — instead, the user might want to retry with a different key, or at a different time maybe.

**Review by v0lkan — APPROVED** [review]
Landing this… the best test env is prod :D 

We need to write integration tests around this too.

File: docs/_pages/0110-configuration.md [hunk 5]
```diff
@@ -465,20 +481,20 @@ should exist, and it should be initiated to a default data as follows before
...
+               # workload’s name.
+               # For example, this is an VSecM-managed secret for `example`
+               # with the SPIFFE ID 
+               # `"spiffe://vsecm.com/workload/example\
+               #  /ns/{{ .PodMeta.Namespace }}\
+               #  /sa/{{ .PodSpec.ServiceAccountName }}\
+               #  /n/{{ .PodMeta.Name }}"`
+               name: vsecm-secret-example
+               namespace: default
+             type: Opaque{% endraw %}
```
Comment (v0lkan) [inline docs/_pages/0110-configuration.md:497]: extra formatting here, can we remove?

**Comment by v0lkan** [comment]
Approved; however, there are merge conflicts.

**Comment by v0lkan** [comment]
I’m intentionally holding this merge (and all other risky merges) until Friday.

**Comment by gurkanguray** [comment]
Closing this pull request due to the many conflicts. The new PR can be found at https://github.com/vmware-tanzu/secrets-manager/pull/725
